File: backend/schemas/portfolio.py
from datetime import date as dt_date
from datetime import datetime

from pydantic import BaseModel, ConfigDict


class PortfolioItemCreate(BaseModel):
    game_id: str | None = None
    model_id: str | None = None
    confidence_level: str | None = None
    bet_type: str | None = None
    stake: float | None = None
    predicted_probability: float | None = None
    ev: float | None = None
    recommendation_stars: int | None = None


class PortfolioCreate(BaseModel):
    date: dt_date | None = None
    total_risk: float | None = None
    expected_roi: float | None = None
    kelly_exposure: float | None = None
    model_id: str | None = None
    items: list[PortfolioItemCreate] = []


class PortfolioRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    date: dt_date | None
    total_risk: float | None
    expected_roi: float | None
    kelly_exposure: float | None
    model_id: str | None
    created_at: datetime


class PortfolioItemRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    portfolio_id: int
    game_id: str | None
    model_id: str | None
    confidence_level: str | None
    bet_type: str | None
    stake: float | None
    predicted_probability: float | None
    ev: float | None
    recommendation_stars: int | None
    created_at: datetime


class BacktestResultCreate(BaseModel):
    model_id: str
    game_id: str
    predicted_probability: float
    actual_outcome: bool | None = None
    edge: float | None = None
    roi: float | None = None


class BacktestResultRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    model_id: str
    game_id: str
    predicted_probability: float | None
    actual_outcome: bool | None
    edge: float | None
    roi: float | None
    created_at: datetime


class ModelEvaluationCreate(BaseModel):
    model_id: str
    accuracy: float | None = None
    calibration_error: float | None = None
    avg_roi: float | None = None
    brier_score: float | None = None
    notes: str | None = None


class ModelEvaluationRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    model_id: str
    evaluated_at: datetime | None
    accuracy: float | None
    calibration_error: float | None
    avg_roi: float | None
    brier_score: float | None
    notes: str | None
    created_at: datetime

File: frontend/src/types/portfolio.ts
export interface ScoreBreakdownEntry {
  value: number | string;
  points: number;
  max: number;
  applied?: string;
}

export interface IntelligenceScoreData {
  score: number;
  stars: number;
  risk_level: string;
  breakdown: Record<string, ScoreBreakdownEntry>;
}

export interface PortfolioItemData {
  id: number;
  portfolio_id?: number;
  game_id: string | null;
  model_id: string | null;
  confidence_level: string | null;
  bet_type: string | null;
  stake: number | null;
  predicted_probability: number | null;
  ev: number | null;
  recommendation_stars: number | null;
}

export interface Portfolio {
  id: number;
  date: string | null;
  total_risk: number | null;
  expected_roi: number | null;
  kelly_exposure: number | null;
  model_id: string | null;
  items: PortfolioItemData[];
}
